src/Moq/Protected/ProtectedMock.cs

```diff
@@ -359,6 +383,193 @@ private static Expression<Action<T>> GetSetterExpression(PropertyInfo property,
+			ThrowIfGetterNotApplicable(indexer);
+
+			return GetIndexerGetterExpression(indexer, indexerKeys);
+		}
+
+		private static LambdaExpression GetIndexerGetterExpression(PropertyInfo property, object[] indexerKeys)
+		{
+			var param = Expression.Parameter(typeof(T), "mock");
+
+			return Expression.Lambda(
+				Expression.Call(param, property.GetGetMethod(true), IndexerArgsToExpressions(indexerKeys, property, true)),
+				param
+			);
+		}
+		
+		private static LambdaExpression GetSetterExpressionIncludingIndexer(string propertyName, object value, bool exact, string propertyNameParameterName, bool throwIfMemberMissing = true)
+		{
+			Guard.NotNullOrEmpty(propertyName, propertyNameParameterName);
+
+			var property = GetPropertyIncludingIndexer(propertyName, true, value, exact);
+			if (property == null && !throwIfMemberMissing)
+			{
+				return null;
+			}
+
+			ThrowIfMemberMissing(propertyName, property);
+			ThrowIfSetterNotApplicable(property);
+
+			Expression<Action<T>> expression;
+			if (property.GetIndexParameters().Length > 0)
+			{
+				expression = GetIndexerSetterExpression(property, CastValueForIndexer(value, propertyName));
+			}
+			else
+			{
+				expression = GetSetterExpression(property, ToExpressionArg(property.PropertyType, value));
+
+			}
+			return expression;
+		}
+
+		private static LambdaExpression GetSetterExpressionIncludingIndexerParams(string propertyName, object[] value, bool exact, string propertyNameParameterName, bool throwIfMemberMissing = true)
+		{
+			Guard.NotNullOrEmpty(propertyName, propertyNameParameterName);
+
+			var property = GetPropertyIncludingIndexer(propertyName, true, value, exact);
+			if (property == null && !throwIfMemberMissing)
+			{
+				return null;
+			}
+
+			ThrowIfMemberMissing(propertyName, property);
+			ThrowIfSetterNotApplicable(property);
+
+			Expression<Action<T>> expression;
+			if (property.GetIndexParameters().Length > 0)
+			{
+				expression = GetIndexerSetterExpression(property, value);
+			}
+			else
+			{
+				expression = GetSetterExpression(property, ToExpressionArg(property.PropertyType, value[0]));
+
+			}
+			return expression;
+		}
+
+		private static LambdaExpression GetGetterExpressionIncludingIndexer<TResult>(string propertyName, object value, bool exact, string propertyNameParameterName)
+		{
+			Guard.NotNullOrEmpty(propertyName, propertyNameParameterName);
+			LambdaExpression expression = null;
+			var property = GetPropertyIncludingIndexer(propertyName, false, value, exact);
+			if (property != null)
+			{
+				ThrowIfGetterNotApplicable(property);
+
+				if (property.GetIndexParameters().Length > 0)
+				{
+					expression = GetIndexerGetterExpression(property, value as object[]);
+				}
+				else
+				{
+					expression = GetMemberAccess<TResult>(property);
+				}
+			}
+
+			return expression;
+		}
+
+		private void GetGetterIncludingIndexerOrMethodExpression<TResult>(string methodOrPropertyName, string methodOrPropertyNameParameterName,
```

Comment:
The `Get…` name prefix suggests that this method returns a value (which isn't the case). Could we rename it to e.g. `For…` or `DoWith…`?

(Same goes for this method's `…Setter…` counterpart.)